radareorg/cutter
# Hexedit

**Detailed description**

This PR replaced the main part of hexdump widget with one that is written from scratch. Thanks @pelijah for writing the initial implementation.

**Test plan (required)**
* basic functionality
  * Shows bytes ✔
  * Scroll to top ✔
  * selection works
    * selecting bytes ✔ 
    * selecting text ✔
    * synchronization between selecting text and bytes ✔
    * select forwards and backwards ✔
    * near the end and start of line ✔
    * shift click ✔
  * keyboard navigation 
    * arrows ✔
    * pageup/down ✔
    * start/end of line ✔
    * selection for all of above ✔
    * near the ends ✔
  * copy bytes ✔ *3
  * copy text ✔
  * copy address (menu and shortcut) ✔
  * changing display mode (each setting at least once + some combinations) 
    * item size ✔
    * format ✔
    * columns ✔ *0
    * hexpairs ✔
    * endian ✔
* right panel
  * selection gets updated when selecting ✔
  * correct range (check using one of byte printers) ✔
  * changing display mode works ✔
* seek synchronization
  * seek in the colored bar at top ✔
  * function list ✔
  * disassm ✔
  * seek in hexdump changes global seek ✔ *1
  * sync/unsync in context menu disables seek synchronization both ways ✔
* setting changes
  * font change ✔
  * color change ✔
* high dpi mode *2


*0 It is possible to force unaligned offset by increasing column count, gets aligned when scrolling - should probably create an issue for adding switch to enable/disable offset alignment. Would be useful for inspecting unaligned multibyte integers, floats.
*1 disasembly widget doesn't like seeking to the middle of instruction - out of scope for this issue
*2 side panel restoration button gets pixelated - should be fixed together with other icons
*3 currently always gets copied as bytes - should create two follow up issues. One for changing default copy behavior to copy exactly as displayed. Second for adding "copy as" menu which allows copying in various formats supported by r2 `p` commands.

![hexwdiget](https://user-images.githubusercontent.com/7101031/57572399-66910180-7422-11e9-9429-08689e9c0385.png)


**Closing issues**

closes #1473, closes #1057, closes #1472, closes #1064, closes #1487, closes #1068 (was already implemented in old hex widget), closes #1494 

~~Doesn't close~~ Keeps #1060 open - implements the drawing mode but doesn't synchronize the setting.
